src/dos/cdrom_image.cpp

```diff
@@ -223,39 +369,130 @@ bool CDROM_Interface_Image::GetMediaTrayStatus(bool& mediaPresent, bool& mediaCh
 	mediaPresent = true;
 	mediaChanged = false;
 	trayOpen = false;
+
+#ifdef DEBUG
+	LOG_MSG("%s CDROM: GetMediaTrayStatus present=%d, changed=%d, open=%d", get_time(), mediaPresent, mediaChanged, trayOpen);
+#endif
+
 	return true;
 }
 
-bool CDROM_Interface_Image::PlayAudioSector(unsigned long start,unsigned long len)
+bool CDROM_Interface_Image::PlayAudioSector(unsigned long start, unsigned long len)
 {
-	// We might want to do some more checks. E.g valid start and length
-	SDL_mutexP(player.mutex);
-	player.cd = this;
-	player.currFrame = start;
-	player.targetFrame = start + len;
-	int track = GetTrack(start) - 1;
-	if(track >= 0 && tracks[track].attr == 0x40) {
+	bool is_playable(false);
+	const int track = GetTrack(start) - 1;
+
+	// The CDROM Red Book standard allows up to 99 tracks, which includes the data track
+	if ( track < 0 || track > 99 )
+		LOG(LOG_MISC, LOG_WARN)("Game tried to load track #%d, which is invalid", track);
+
+	// Attempting to play zero sectors is a no-op
+	else if (len == 0)
+		LOG(LOG_MISC, LOG_WARN)("Game tried to play zero sectors, skipping");
+
+	// The maximum storage achieved on a CDROM was ~900MB or just under 100 minutes
+	// with overburning, so use this threshold to sanity-check the start sector.
+	else if (start > 450000)
+		LOG(LOG_MISC, LOG_WARN)("Game tried to read sector %lu, which is beyond the 100-minute maximum of a CDROM", start);
+
+	// We can't play audio from a data track (as it would result in garbage/static)
+	else if(track >= 0 && tracks[track].attr == 0x40)
 		LOG(LOG_MISC,LOG_WARN)("Game tries to play the data track. Not doing this");
-		player.isPlaying = false;
-		//Unclear wether return false should be here. 
-		//specs say that this function returns at once and games should check the status wether the audio is actually playing
-		//Real drives either fail or succeed as well
-	} else player.isPlaying = true;
-	player.isPaused = false;
-	SDL_mutexV(player.mutex);
-	return true;
+
+	// Checks passed, setup the audio stream
+	else {
+		TrackFile* trackFile = tracks[track].file;
+
+		// Convert the playback start sector to a time offset (milliseconds) relative to the track
+		const Bit32u offset = tracks[track].skip + (start - tracks[track].start) * tracks[track].sectorSize;
+		is_playable = trackFile->seek(offset);
+
+		// only initialize the player elements if our track is playable
+		if (is_playable) {
+			const Bit8u channels = trackFile->getChannels();
+			const Bit32u rate = trackFile->getRate();
+
+			player.cd = this;
+			player.trackFile = trackFile;
+			player.startFrame = start;
+			player.currFrame = start;
+			player.numFrames = len;
+			player.bufferPos = 0;
+			player.bufferConsumed = 0;
+			player.isPlaying = true;
+			player.isPaused = false;
+
+
+#if defined(WORDS_BIGENDIAN)
+			if (trackFile->getEndian() != AUDIO_S16SYS)
+#else
+			if (trackFile->getEndian() == AUDIO_S16SYS)
+#endif
+				player.addSamples = channels ==  2  ? &MixerChannel::AddSamples_s16 \
```

Comment:
Callback to my earlier comment about getEndian() ^^^

Reply:
The preprocessors are now gone in cdda-merge-5; although I'm still using getEndian to assess each track.